core/shared/src/main/scala/monocle/function/At.scala

```diff
@@ -13,27 +13,18 @@ import scala.collection.immutable.{ListMap, SortedMap}
 @implicitNotFound(
   "Could not find an instance of At[${S},${I},${A}], please check Monocle instance location policy to " + "find out which import is necessary"
 )
-abstract class At[S, I, A] extends Serializable {
-  def at(i: I): Lens[S, A]
-}
-
-abstract class AtSingleton[S, I <: Singleton, A] extends At[S, I, A] {
+abstract class At[S, -I, A] extends Serializable {
   def at(i: I): Lens[S, A]
 }
 
 trait AtFunctions {
-  def at[S, I, A](i: I)(implicit ev: At[S, I, A]): Lens[S, A] = ev.at(i)
+  def at[S, I, A](i: I)(implicit ev: At[S, i.type, A]): Lens[S, A] = ev.at(i)
```

Comment:
oh, that's interesting. If you have time, would you mind to explain me why you use `i.type` here?

Reply:
For primitive types it has the same effect as `<: Singleton` (prevents widening) and for reference types it's exactly what it looks like - the singleton type of the parameter. But the cool thing is that it works in both cases with the same method. Only we need `I` to be contravariant so that `At[S, i.type, A]` is also eligible for `At[S, I, A]`.

Reply:
ah I see, thank you!